File: src/circuitBreaker.js
// Implementing half-open state logic and metrics tracking for circuit breaker

class CircuitBreaker {
    constructor() {
        this.state = 'CLOSED';
        this.failureCount = 0;
        this.successCount = 0;
        this.metrics = [];
    }

    // Method to simulate failure
    fail() {
        this.failureCount++;
        this.metrics.push({ state: this.state, success: false });
        if (this.state === 'CLOSED') {
            this.state = 'OPEN';
        }
    }

    // Method to simulate success
    succeed() {
        this.successCount++;
        this.metrics.push({ state: this.state, success: true });
        if (this.state === 'OPEN') {
            this.state = 'HALF-OPEN';
        }
    }

    // Method to reset state
    reset() {
        this.state = 'CLOSED';
        this.failureCount = 0;
        this.successCount = 0;
        this.metrics = [];
    }

    // Method to get metrics
    getMetrics() {
        return this.metrics;
    }
}

File: circuit_breaker.java
// Improved Circuit Breaker implementation with Half-Open State and Detailed Metrics

public class CircuitBreaker {
    private enum State { CLOSED, OPEN, HALF_OPEN }
    private State state;
    private int failureCount;
    private int successCount;
    private long lastFailureTime;
    private long timeout;

    public CircuitBreaker(long timeout) {
        this.state = State.CLOSED;
        this.timeout = timeout;
        this.failureCount = 0;
        this.successCount = 0;
    }

    public void call() {
        switch (state) {
            case CLOSED:
                // Execute call and handle success/failure
                break;
            case OPEN:
                // Reject call
                break;
            case HALF_OPEN:
                // Allow limited calls
                break;
        }
    }

    private void onSuccess() {
        successCount++;
        if (state == State.HALF_OPEN) {
            state = State.CLOSED;
        }
    }

    private void onFailure() {
        failureCount++;
        lastFailureTime = System.currentTimeMillis();
        if (failureCount >= threshold) {
            state = State.OPEN;
        }
    }

    public String getMetrics() {
        return "Failures: " + failureCount + ", Successes: " + successCount;
    }
}